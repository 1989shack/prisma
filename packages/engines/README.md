# `@prisma/engines`

This package is used to download the Prisma Engines, namely the Query Engine, Migration Engine, Introspection Engine and Prisma Format.

Its version Before Prisma 4 had a one to one mapping to the engine hash in its version.
Since Prisma 4 its version follows the `prisma/prisma` monorepo versioning and it uses the `@prisma/engines-version` package to map to the engines hash.

In the `postinstall` hook of this package, all engines available for the current platform are downloaded from the Prisma CDN `https://binaries.prisma.sh`.

You should probably not use this package directly, but instead use one of these:

- [`prisma` CLI](https://www.npmjs.com/package/prisma)
- [`@prisma/client`](https://www.npmjs.com/package/@prisma/client)
