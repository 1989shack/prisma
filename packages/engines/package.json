{
  "name": "@prisma/engines",
  "version": "0.0.0",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "license": "Apache-2.0",
  "author": "Tim Suchanek <suchanek@prisma.io>",
  "devDependencies": {
    "@prisma/debug": "workspace:*",
    "@prisma/engines-version": "3.16.0-46.7913709f1e48160e65e366f365c239e3017a9c2e",
    "@prisma/fetch-engine": "workspace:*",
    "@prisma/get-platform": "workspace:*",
    "@swc/core": "1.2.197",
    "@swc/jest": "0.2.21",
    "@types/jest": "28.1.1",
    "@types/node": "16.11.39",
    "execa": "5.1.1",
    "jest": "27.5.1",
    "typescript": "4.7.3"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/prisma/prisma.git",
    "directory": "packages/engines"
  },
  "scripts": {
    "dev": "DEV=true node -r esbuild-register helpers/build.ts",
    "build": "node -r esbuild-register helpers/build.ts",
    "test": "jest",
    "prepublishOnly": "pnpm run build",
    "postinstall": "node scripts/postinstall.js"
  },
  "files": [
    "dist",
    "download",
    "scripts"
  ],
  "sideEffects": false
}
