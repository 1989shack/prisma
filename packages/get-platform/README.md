# @prisma/get-platform

Platform detection.

## Usage

```ts
import { getPlatform } from '@prisma/get-platform'

const platform = await getPlatform()
```
