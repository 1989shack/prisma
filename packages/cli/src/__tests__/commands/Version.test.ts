import { getCliQueryEngineType } from '@prisma/engines'
import { download, EngineType } from '@prisma/fetch-engine'
import { getPlatform } from '@prisma/get-platform'
import { engineEnvVarMap, jestConsoleContext, jestContext } from '@prisma/sdk'
import makeDir from 'make-dir'
import path from 'path'

const ctx = jestContext.new().add(jestConsoleContext()).assemble()
const testIf = (condition: boolean) => (condition ? test : test.skip)
const useNodeAPI = getCliQueryEngineType() === EngineType.libqueryEngine
const version = '5a2e5869b69a983e279380ec68596b71beae9eff'

describe('version', () => {
  // Node-API Tests

  testIf(useNodeAPI)('basic version (Node-API)', async () => {
    const data = await ctx.cli('--version')
    expect(cleanSnapshot(data.stdout)).toMatchSnapshot()
  })

  testIf(useNodeAPI)(
    'version with custom engines (Node-API)',
    async () => {
      const enginesDir = path.join(__dirname, 'version-test-engines')
      await makeDir(enginesDir)
      const enginesPaths = await download({
        engines: {
          'introspection-engine': enginesDir,
          'migration-engine': enginesDir,
          'prisma-fmt': enginesDir,
          'libquery-engine': enginesDir,
        },
        version,
        failSilent: false,
      })
      // This Omits query-engine from the map
      const { ['query-engine']: qe, ...envVarMap } = engineEnvVarMap

      const platform = await getPlatform()

      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        process.env[envVar] = enginesPaths[engine][platform]
        // console.debug(`Setting ${envVar} to ${enginesPaths[engine][platform]}`)
      }

      const data = await ctx.cli('--version')
      expect(cleanSnapshot(data.stdout)).toMatchSnapshot()

      // cleanup
      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        delete process[envVar]
      }
    },
    50000,
  )

  // Binary Tests

  testIf(!useNodeAPI)('basic version', async () => {
    const data = await ctx.cli('--version')
    expect(cleanSnapshot(data.stdout)).toMatchSnapshot()
  })

  testIf(!useNodeAPI)(
    'version with custom engines',
    async () => {
      const enginesDir = path.join(__dirname, 'version-test-engines')
      await makeDir(enginesDir)
      const enginePaths = await download({
        engines: {
          'introspection-engine': enginesDir,
          'migration-engine': enginesDir,
          'prisma-fmt': enginesDir,
          'query-engine': enginesDir,
        },
        version,
        failSilent: false,
      })

      const platform = await getPlatform()
      const { ['libquery-engine']: qe, ...envVarMap } = engineEnvVarMap
      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        process.env[envVar] = enginePaths[engine][platform]
        // console.debug(`Setting ${envVar} to ${enginePaths[engine][platform]}`)
      }

      const data = await ctx.cli('--version')
      expect(cleanSnapshot(data.stdout)).toMatchSnapshot()

      // cleanup
      for (const engine in envVarMap) {
        const envVar = envVarMap[engine]
        delete process[envVar]
      }
    },
    50000,
  )
})

function cleanSnapshot(str: string): string {
  //return str.replace(/:(.*)/g, ': placeholder')

  // sanitize engine path
  // Query Engine (Node-API) : libquery-engine e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at ../../home/runner/work/prisma/prisma/node_modules/.pnpm/@prisma+engines@3.11.0-41.e996df5d66a2314d1da15d31047f9777fc2fbdd9/node_modules/@prisma/engines/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node)
  // =>
  // Query Engine (Node-API) : libquery-engine e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at sanitized_path/node_modules/@prisma/engines/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node)
  str.replace(/at (.*)\/node_modules\/@prisma\/engines/g, 'at sanitized_path/node_modules/@prisma/engines')

  return str
}
