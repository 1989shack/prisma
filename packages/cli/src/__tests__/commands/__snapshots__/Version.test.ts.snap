// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`version basic (Node-API) 1`] = `
prisma                  : placeholder
@prisma/client          : placeholder
Current platform        : placeholder
Query Engine (Node-API) : libquery-engine e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at sanitized_path/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node)
Migration Engine        : migration-engine-cli e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at sanitized_path/migration-engine-TEST_PLATFORM)
Introspection Engine    : introspection-core e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at sanitized_path/introspection-engine-TEST_PLATFORM)
Format Engine           : prisma-fmt e996df5d66a2314d1da15d31047f9777fc2fbdd9 (at sanitized_path/prisma-fmt-TEST_PLATFORM)
Default Engines Hash    : placeholder
Studio                  : placeholder
`;

exports[`version basic (binary) 1`] = `
prisma                : placeholder
@prisma/client        : placeholder
Current platform      : placeholder
Query Engine (Binary) : placeholder
Migration Engine      : placeholder
Introspection Engine  : placeholder
Format Engine         : placeholder
Default Engines Hash  : placeholder
Studio                : placeholder
`;

exports[`version with custom engines (Node-API) 1`] = `
prisma                  : placeholder
@prisma/client          : placeholder
Current platform        : placeholder
Query Engine (Node-API) : libquery-engine 5a2e5869b69a983e279380ec68596b71beae9eff (at sanitized_path/libquery_engine-TEST_PLATFORM.LIBRARY_TYPE.node, resolved by PRISMA_QUERY_ENGINE_LIBRARY)
Migration Engine        : migration-engine-cli 5a2e5869b69a983e279380ec68596b71beae9eff (at sanitized_path/migration-engine-TEST_PLATFORM, resolved by PRISMA_MIGRATION_ENGINE_BINARY)
Introspection Engine    : introspection-core 5a2e5869b69a983e279380ec68596b71beae9eff (at sanitized_path/introspection-engine-TEST_PLATFORM, resolved by PRISMA_INTROSPECTION_ENGINE_BINARY)
Format Engine           : prisma-fmt 5a2e5869b69a983e279380ec68596b71beae9eff (at sanitized_path/prisma-fmt-TEST_PLATFORM, resolved by PRISMA_FMT_BINARY)
Default Engines Hash    : placeholder
Studio                  : placeholder
`;

exports[`version with custom engines (binary) 1`] = `
prisma                : placeholder
@prisma/client        : placeholder
Current platform      : placeholder
Query Engine (Binary) : placeholder
Migration Engine      : placeholder
Introspection Engine  : placeholder
Format Engine         : placeholder
Default Engines Hash  : placeholder
Studio                : placeholder
`;
